engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java

```diff
@@ -114,112 +74,26 @@ public void processRecord(final TypedRecord<ProcessInstanceMigrationRecord> comm
     final long processInstanceKey = value.getProcessInstanceKey();
     final long targetProcessDefinitionKey = value.getTargetProcessDefinitionKey();
     final var mappingInstructions = value.getMappingInstructions();
-
     final ElementInstance processInstance = elementInstanceState.getInstance(processInstanceKey);
 
-    if (processInstance == null) {
-      final String reason =
-          String.format(ERROR_MESSAGE_PROCESS_INSTANCE_NOT_FOUND, processInstanceKey);
-      responseWriter.writeRejectionOnCommand(command, RejectionType.NOT_FOUND, reason);
-      rejectionWriter.appendRejection(command, RejectionType.NOT_FOUND, reason);
-      return;
-    }
-
-    final boolean isTenantAuthorized =
-        TenantAuthorizationCheckerImpl.fromAuthorizationMap(command.getAuthorizations())
-            .isAuthorized(processInstance.getValue().getTenantId());
-    if (!isTenantAuthorized) {
-      final String reason =
-          String.format(ERROR_MESSAGE_PROCESS_INSTANCE_NOT_FOUND, processInstanceKey);
-      responseWriter.writeRejectionOnCommand(command, RejectionType.NOT_FOUND, reason);
-      rejectionWriter.appendRejection(command, RejectionType.NOT_FOUND, reason);
-      return;
-    }
-
-    if (processInstance.getValue().getParentProcessInstanceKey() != NO_PARENT) {
-      throw new ChildProcessMigrationException(processInstanceKey);
-    }
+    requireNonNullProcessInstance(processInstance, processInstanceKey);
+    requireAuthorizedTenant(
+        command.getAuthorizations(), processInstance.getValue().getTenantId(), processInstanceKey);
+    requireNullParent(processInstance.getValue().getParentProcessInstanceKey(), processInstanceKey);
+    requireNonDuplicateSourceElementIds(mappingInstructions, processInstanceKey);
```

Comment:
@korthout Nice suggestion. 👍 The delegation of complexity out of the `processRecord()` method will provide better focus on migration logic. I would agree with any other idea that delegates complexity out of migration processor because the processor logic is expected to be complex by itself.

I updated the code based on your suggestion. However, by looking at result, I am not sure if the complexity is actually eliminated. Because of those reasons:
* As you mentioned, now precondition checker keeps state and we should skip setting values to it.
* Preconditions are now called through object `checker.require*`. That is not a common usage of preconditions such that they are usually static methods e.g. `assertThat()` or Guava's preconditions `Preconditions.checkArgument()`.

I updated the code. Please let me know what you think. 👀